ebib-ivy.el

```diff
@@ -0,0 +1,48 @@
+;;; ebib-ivy.el --- Part of Ebib, a BibTeX database manager  -*- lexical-binding: t -*-
+
+;; Copyright (c) 2021 Hugo Heagren
+;; All rights reserved.
+
+;; Author: Hugo Heagren <hugo@heagren.com>
+;; Maintainer: Joost Kremers <joostkremers@fastmail.fm>
+
+;; Redistribution and use in source and binary forms, with or without
+;; modification, are permitted provided that the following conditions
+;; are met:
+;;
+;; 1. Redistributions of source code must retain the above copyright
+;;    notice, this list of conditions and the following disclaimer.
+;; 2. Redistributions in binary form must reproduce the above copyright
+;;    notice, this list of conditions and the following disclaimer in the
+;;    documentation and/or other materials provided with the distribution.
+;; 3. The name of the author may not be used to endorse or promote products
+;;    derived from this software without specific prior written permission.
+;;
+;; THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+;; IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+;; OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+;; IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+;; INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+;; NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES ; LOSS OF USE,
+;; DATA, OR PROFITS ; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+;; THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+;; (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+;; THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+;;; Commentary:
+
+;; This file is part of Ebib, a BibTeX database manager for Emacs.  It contains
+;; the code to integrate ivy's features more tightly into the interface.
+;; To use this integration, add (require 'ebib-ivy) to init.el file.
+
+;;; Code:
+
+(require 'ebib)
```

Comment:
I'm a bit ambivalent about this `require`. I guess it avoids a bunch of warnings in `ebib-ivy.el` and it won't do any harm, but still...

I do suspect a `(require ivy)` will be necessary, BTW.

Reply:
> I'm a bit ambivalent about this require. I guess it avoids a bunch of warnings in ebib-ivy.el and it won't do any harm, but still...

Can I ask why? I'm pretty new to elisp and emacs packaging, so I don't know about the vagueries of this sort of thing. My reasoning was that ivy-ebib will require various (not autoloaded) functions from ebib (like `ebib-copy-field-contents`), though the reverse will never be true, so there will never be a requirement circle. I can take it out though if you think that would be better. 

> I do suspect a (require ivy) will be necessary, BTW.

Ah yes, I think so!

Reply:
> > I'm a bit ambivalent about this require. I guess it avoids a bunch of warnings in ebib-ivy.el and it won't do any harm, but still...
> 
> Can I ask why?

Sure. :slightly_smiling_face: 

> I'm pretty new to elisp and emacs packaging, so I don't know about the vagueries of this sort of thing. My reasoning was that ivy-ebib will require various (not autoloaded) functions from ebib (like `ebib-copy-field-contents`), though the reverse will never be true, so there will never be a requirement circle.

It's really more of a philosophical thing... To me, `ebib-ivy.el` is an optional dependency of `ebib`, not the other way around, so `require` doesn't feel "right".

Though you could also argue that since `ebib.el` doesn't actually use any code from `ebib-ivy.el`, it's more of an add-on rather than a dependency, and then a case could be made that `require` is appropriate.

And you're right, it doesn't lead to any practical problems.

>  I can take it out though if you think that would be better.

Just leave it in. I may at some point decide to replace the `require` with a set of `defvar`s and/or `declace-function`s as I did in `org-ebib.el`. Or I may decide to put a `(require 'ebib)` in `org-ebib.el` and be done with it. :laughing: